GOVERNANCE_DOC.md

```diff
@@ -0,0 +1,597 @@
+
+- **Proposal submission:** Proposal is submitted to the blockchain with a deposit 
+- **Vote:** Once deposit reaches a certain value (`MinDeposit`), proposal is confirmed and vote opens. Bonded Atom holders can then send `TxVote` transactions to vote on the proposal
+- If the proposal involves a software upgrade
+  - **Signal:** Validator start signaling that they are ready to switch to the new version
+  - **Switch:** Once more than 2/3rd validators have signaled their readiness to switch, their software automatically flips to the new version
+  
+## Proposal submission
+
+### Right to submit a proposal
+
+Any Atom holder, whether bonded or unbonded, can submit proposals by sending a `TxProposal` transaction. Once a proposal is submitted, it is identified by its unique `proposalID`.
+
+### Proposal filter (minimum deposit)
+
+To prevent spam, proposals must be submitted with a deposit in Atoms. Voting period will not start as long as the proposal's deposit is smaller than the minimum deposit parameter `MinDeposit`.
+
+When a proposal is submitted, it has to be accompagnied by a deposit that must be strictly positive but that can be inferior to `MinDeposit`. Indeed, the submitter need not pay for the entire deposit on its own. If a proposal's deposit is strictly inferior to `MinDeposit`, other Atom holders can increase the proposal's deposit by sending a `TxDeposit` transaction. Once the proposals's deposit reaches `minDeposit`, it enters voting period. 
+
+### Deposit refund
+
+There are two instances where Atom holders that deposited can claim back their deposit:
+- If the proposal is accepted
+- If the proposal's deposit does not reach `MinDeposit` for a period longer than `mMxDepositPeriod` (initial value: 2 months). Then the proposal is considered closed and nobody can deposit on it anymore.
+
+In such instances, Atom holders that deposited can send a `TxClaimDeposit` transaction to retrieve their share of the deposit.
+
+### Proposal types
+
+In the initial version of the governance module, there are two types of proposal:
+- `PlainTextProposal`. All the proposals that do not involve a modification of the source code go under this type. For example, an opinion poll would use a proposal of type `PlainTextProposal`
+- `SoftwareUpgradeProposal`. If accepted, validators are expected to update their software in accordance with the proposal. They must do so by following a 2-steps process described in the [Software Upgrade](#software-upgrade) section below. Software upgrade roadmap may be discussed and agreed on via `PlainTextProposals`, but actual software upgrades must be performed via `SoftwareUpgradeProposals`.
+
+### Proposal categories
+
+There are two categories of proposal:
+- `Regular`
+- `Urgent`
+
+These two categories are strictly identical except that `Urgent` proposals can be accepted faster if a certain condition is met. For more information, see [Threshold](#threshold) section.
```

Comment:
I don't think we need the urgent message type... If it's urgent the validators can just push it through on their own accord without governance and then create a vote post-mortum, this non-governance mechanism effectively covers urgent changes @jaekwon What do you think about urgent proposals?

Reply:
Personally I think it's good to have a public and open process, so that decisions have legitimacy. As far as I remember jae thought the same. It's not that complicated to implement.

Reply:
I think flagging a proposal as "urgent" is fine, because it can wake up validators, red flashing lights and sirens, paging on-call devops etc.  If it wasn't actually urgent, then we'll maybe figure out some punishment for it later.  The whitepaper specifies or hints at the ability to say "burn that deposit, this proposal is junk", and that would also work for "burn that deposit, this wasn't urgent".

Otherwise, proposals should work the same whether urgent or not.  It's not considered having "passed" unless it receives >50% of the votes, including delegation power.  If the delegators are too late in showing up for their vote, then perhaps it was (actually) an urgent matter, and it passed despite the lack of delegators' support.

If we require a waiting period before a proposal is decided to have "passed" in order to let the delegators voice their opinions, then we lose the opportunity to respond to urgent matters.  So there must be an implicit contract that for urgent matters (regardless of the "urgent" flag), the validators have to make a best-case judgement call.  Perhaps, how to resolve the delegator-lag issue should be specified in the proposal, so that validators can judge for themselves what they're actually voting for.  Until we see some options it seems to me that it's better to leave this open for now.

In any case, whether a proposal is marked "urgent" or not should be cosmetic, and have no effect on the process of actually voting or deciding.

For example, a Byzantine proposer can propose the vote for a really urgent matter, but not flag the proposal as "urgent".  Now, should voters vote yes for that, or vote yes for the alternative (newer) proposal that *is* marked "urgent"?  The solution to this conundrum/wrinkle-in-logic is to ignore it.  But it could still be useful to wake up devops in the middle of the night.

Reply:
Since the existence of an "urgent" flag may confuse participants about the true urgency of proposals, maybe it's better to leave it out, and have a convention of including the "URGENT" text in the proposal (or in the future, comments).  I dunno.

Reply:
hmm yeah interesting about the ambiguity of use of the urgent flag... Maybe the urgentness of a proposal could be determined not by the person who submits the proposal but by the people that are voting for the proposal. If people begin to consitently vote a propsal as urgent then it would automatically take on the urgent type... I think this could avoid some of the problems outlined... but it's also a bit more confusing..

I guess my comment is we still need an out of band process for validators to change sensistive issues as well such as the example with the Monero upgrade in the past... this can be dealt with seperately however, and is maybe an edge case outside of the standard "urgent" proposal being outlined... but then again maybe not? I feel like the majority "urgent" proposals will be security patches and should be sensitive as to not alert potential malicious actors who could take advantage of a bug before it's patched... am I wrong? what other types of urgent proposals can you guys think of which would really merit not going out of protocol and not waiting two weeks??? @jaekwon @gamarin2

Reply:
I think if such a thing were to happen, it would happen off-chain, and in any case it doesn't sound like something that would help us pass the Howie test (e.g. see SEC analysis of TheDAO and need for transparent decentralized coordination).

I think we could maybe propose mechanisms to adopt some off-chain policy, but I think it's something to do after the chain has launched, if ever.  I'm not comfortable using a chain where patches can happen in secret if it implies that there aren't enough eyeballs in the PR process.  It could work for other projects but maybe we can do something different.  If we have enough eyeballs on Tendermint/Cosmos, would all serious bugs be discovered with sufficient time for review?

Reply:
https://en.wikipedia.org/wiki/Linus%27s_Law Lets run this way.

Reply:
So in conclusion:
- no urgent flag as it leads to ambiguity
- no min voting period. once a proposal is accepted by the majority it is considered passed

I think in the documentation we should highlight to validators though, that it is probably a reasonable thing to do to give their delegators 2 weeks to vote on things.

Reply:
Personally I agree with having a min voting period. Otherwise delegators will lose a lot of control. Maybe they will switch delegation if their validator always votes immediately, but I think it would be cleaner to just have a min voting period.

And having a min voting period makes an urgent flag somewhat necessary, under the assumption that we want to be able to upgrade things in less than the min voting period. If we have an urgent flag though, the validators might also abuse it to just pass all proposals without delegator interaction.

I think the question is:
Do we rather want delegator ability to influence votes more heavily or pass proposals faster?

I think it's a tradeoff we have to decide on, or postpone for now and let the community decide through governance later on.
If we decide to leave it to the community I would suggest that we implement min voting period and no urgent proposals.

Reply:
> aren't enough eyeballs in the PR process

very fair point

@adrianbrink I think there was some miscommunication, there are some valid reasons as to why we want the voting period even after a "pass" because delegators can change the outcome of a result by moving their vote to something else from the default (which is whatever their validator voted for)

Reply:
> And having a min voting period makes an urgent flag somewhat necessary, under the assumption that we want to be able to upgrade things in less than the min voting period. If we have an urgent flag though, the validators might also abuse it to just pass all proposals without delegator interaction.

If it happens then delegators will move away from these validators. I really think the debate lies here (c/cing from above): 

- Having it implicit that validators have to leave some time for their delegators to vote if they don't deem the proposal urgent (one process)
- Having it explicit that validators have to leave some time for their delegators to vote if they don't deem the proposal urgent (two processes)

Reply:
> Whenever you are voting something as "urgent" it's only your personal stake which counts towards making the proposal urgent, so if the network has generally low-self bond ratios for validators a proposal could not simply be pushed through by the validators because it's only their small stake which would count towards calling the proposal urgent... the network would then need to wait for some threshold of the delegators to agree that the proposal is urgent befor it would take off before the end of the voting period 🥇 I think this solves the issue

@rigelrozanski. The reason why @jaekwon wanted urgent proposals AFAIK is that sometimes validators have to step up and take action quickly. It's precisely when proposals are urgent that we need inheritance the most so that we can have quick resolution. Delegators will probably be much less reactive and sometimes we won't be able to wait until they wake up.

In the end, I maintain that this is an issue of clarity. With two processes, users have a way to protect themselves from validators, who would have to be actively Byzantine and censor votes OR re-submit the proposal as urgent if they wanted to decide without letting delegators participate. We can't really do much about the former, but the latter gives more clarity to users. If we only have one process, it's way harder for delegators to know if validators are depriving them of voting time.

Plus one process introduces timing issues. A validator with significant voting power could monitor the proposal and time its vote so that it closes the proposal. For example, a validator could close a proposal by voting on day 4, and only on this very day. The same vote any other day would not have closed the vote. I think it's good for users and delegators to know that there is a governance process where this has no chance of happening.

Reply:
You're missing a point here, under two processes there is still an issue with delegators BEING at risk of large validators sensoring them if the large validator submits an urgent proposal... I proposed a solution which addresses this under a one process: 
>SO HERE IS THE SOLUTION: Whenever you are voting something as "urgent" it's only your personal stake which counts towards making the proposal urgent, so if the network has generally low-self bond ratios for validators a proposal could not simply be pushed through by the validators because it's only their small stake which would count towards calling the proposal urgent... the network would then need to wait for some threshold of the delegators to agree that the proposal is urgent before it would take off before the end of the voting period

But yes there is some inherent tradeoff between voting time and proposal urgency, I think that It shouldn't be a problem for the network to coordinate quickly under and urgent proposal and push it to through with an urgent vote using the above mechanism. OR if something is super extremely tremendously important for the decision to push through out of band

Reply:
> @gamarin2 You're missing a point here, under two processes there is still an issue with delegators BEING at risk of large validators sensoring them if the large validator submits an urgent proposal...

No, please re-read my previous post, I've considered this point, cf

> OR re-submit the proposal as urgent if they wanted to decide without letting delegators participate

We can't do anything about that. However, two processes allow us to have more clarity about what's going on. If there are two processes, it will be clear to everyone that someone re-submitted a regular proposal as urgent and, if it is accepted as urgent but not as regular, then we clearly know there is a divergence between validators and delegators. Then, delegators can identify which validators are trying to bypass them (by voting Yes on the urgent proposal) and move away from them. Again, it's a question of clarity.

And if we have inheritance and validator's punishment, it's because we want validators to be active and step up when needed. So it wouldn't make sense to me to deprive them from their delegated voting power when proposals are urgent.

Reply:
are you referring to a comment like this?
> I still think the two processes option is the better because it's more clear and actually less ambiguous. If they don't deem the proposal urgent, validators could vote early without the fear of shutting down the vote and making their delegators unhappy. Likewise, delegators could vote late without the fear of being too late because some validator with sufficient voting power shut down the vote for whatever personal motive.

I agree with the logic, but I still think it's confusing as heck and suspect that it may be difficult to uptake. Two processes has less clarity, think about being a delegator and seeing two identical proposals one marked as urgent and one as not urgent, totally bonkers - I don't think most people will even know how to act under this situation, it's non-intuitive. 

If we do something good with the UI or have explicit proposals which are "non-urgent" and their goal is to overturn an existing "urgent" proposal which has just gone through, then yes it would achieve the same effect - I still see this as a poor choice however... 

why can't the validators just go out of band if a proposal is so so urgent that they can't wait for some basic loose coordination of the network (delegators and validators) to vote that YES a proposal is in fact urgent and should be passed in less than 2 weeks... This is probably very rare situations anyways.

Reply:
> I agree with the logic, but I still think it's confusing as heck and suspect that it may be difficult to uptake. Two processes has less clarity, think about being a delegator and seeing two identical proposals one marked as urgent and one as not urgent, totally bonkers - I don't think most people will even know how to act under this situation, it's non-intuitive.

This problem exists even if we have only one process. Someone can submit the same proposal twice. If we have two processes and what you say happens, then I think it's a good way of finding out if validators and delegators converge in opinion, which is the only stable state of the network. 

> why can't the validators just go out of band if a proposal is so so urgent that they can't wait for some basic loose coordination of the network (delegators and validators) to vote that YES a proposal is in fact urgent and should be passed in less than 2 weeks... This is probably very rare situations anyways.
 
I don't think we can cover every situation. We are just trying to find the best way to proceed for, say, 90% of proposals. For example, if there is a bug that affects the governance code itself, then validators will have no choice but to coordinate off-chain. But these are extreme case. I believe many proposals will be just fine going through the regular process and full voting period. For these proposals it's good that validators and delegators can vote whenever they want without fearing timing issues and power imbalance.

Reply:
> @gamarin2: This problem exists even if we have only one process. Someone can submit the same proposal twice. If we have two processes and what you say happens, then I think it's a good way of finding out if validators and delegators converge in opinion, which is the only stable state of the network.

If there is only 1 process, then there's less need for duplicate proposals.  The beauty of the 1 process system is that whether a proposal is urgent is decided by the voters themselves.